File: src/pi/modules/control_tasks/sensor_control.py
import time
from modules.mcl.flag import Flag
from modules.lib.kalman import Kalman
from modules.lib.helpers import enqueue
from modules.mcl.registry import Registry
from modules.lib.packet import Log, LogPriority
from modules.lib.enums import SensorType, SensorLocation, SensorStatus

class SensorControl():
    def __init__(self, registry: Registry, flag: Flag):
        print("Sensor control")
        self.registry = registry
        self.flag = flag


    def begin(self, config: dict):
        self.config = config
        sensor_config = config["sensors"]["list"]

        self.sensors = {}
        self.boundaries = {}
        self.kalman_args = {}
        for sensor_type in sensor_config:
            self.sensors[sensor_type] = []
            self.boundaries[sensor_type] = {}
            self.kalman_args[sensor_type] = {}
            for sensor_loc in sensor_config[sensor_type]:
                self.sensors[sensor_type].append(sensor_loc)
                self.boundaries[sensor_type][sensor_loc] = sensor_config[sensor_type][sensor_loc]["boundaries"]["waiting"]
                self.kalman_args[sensor_type][sensor_loc] =  sensor_config[sensor_type][sensor_loc]["kalman_args"]
        self.send_interval = self.config["sensors"]["send_interval"]
        self.last_send_time = None
        self.init_kalman(config)
    

    def init_kalman(self, config: dict):
        self.kalman_filters = {}
        for sensor_type in self.sensors:
            self.kalman_filters[sensor_type] = {}
            for sensor_location in self.sensors[sensor_type]:
                args = self.kalman_args[sensor_type][sensor_location]
                self.kalman_filters[sensor_type][sensor_location] = Kalman(args["process_variance"],
                                                                           args["measurement_variance"],
                                                                           args["kalman_value"])


    # Test to make sure sensor values aren't outside the boundaries set in the config. If they are, update the registry with the appropriate SensorStatus.
    def boundary_check(self):
        crits = []
        for sensor_type in self.sensors:
            for sensor_location in self.sensors[sensor_type]:
                _, val, _ = self.registry.get(("sensor_measured", sensor_type, sensor_location))
                kalman_val = self.kalman_filters[sensor_type][sensor_location].update_kalman(val)
                boundaries = self.boundaries[sensor_type][sensor_location]
                self.registry.put(("sensor_normalized", sensor_type, sensor_location), kalman_val)
                if boundaries["safe"][0] <= kalman_val <= boundaries["safe"][1]:
                    self.registry.put(("sensor_status", sensor_type, sensor_location), SensorStatus.SAFE)
                elif boundaries["warn"][0] <= kalman_val <= boundaries["warn"][1]:
                    self.registry.put(("sensor_status", sensor_type, sensor_location), SensorStatus.WARNING)
                else:
                    self.registry.put(("sensor_status", sensor_type, sensor_location), SensorStatus.CRITICAL)
                    crits.append([sensor_type, sensor_location])

        soft = self.registry.get(("general", "soft_abort"))[1]
        # if not hard:
        #     if len(crits) == 0 and soft:
        #         # TODO: Put hard_abort and soft_abort in flag?
        #         # Undo soft abort (since all sensors are back to normal)
        #         self.registry.put(("general", "soft_abort"), False)
        #         log = Log(header="response", message={"header": "Undoing soft abort", "Description": "All sensors have returned to non-critical levels"})
        #         enqueue(self.flag, log, LogPriority.CRIT)
        #         enqueue(self.flag, Log(header="mode", message={"mode": "Normal"}), LogPriority.CRIT)
        
        # executes soft abort
        if len(crits) > 0 and not soft:
                # soft abort if sensor status is critical and send info to GS
                self.registry.put(("general", "soft_abort"), True)
                log = Log(header="response", message={"header": "Soft abort", "Description": sensor_type + " in " + sensor_location + " reached critical levels"})
                enqueue(self.flag, log, LogPriority.CRIT)
                enqueue(self.flag, Log(header="mode", message={"mode": "soft abort"}), LogPriority.CRIT)

    def send_sensor_data(self):
        message = {}
        for sensor_type in self.sensors:
            message[sensor_type] = {}
            for sensor_location in self.sensors[sensor_type]:
                _, val, _ = self.registry.get(("sensor_measured", sensor_type, sensor_location))
                _, kalman_val, _ = self.registry.get(("sensor_normalized", sensor_type, sensor_location))
                _, status, _ = self.registry.get(("sensor_status", sensor_type, sensor_location))                
                message[sensor_type][sensor_location] = {"value": (val, kalman_val), "status": status}
        log = Log(header="sensor_data", message=message)
        enqueue(self.flag, log, LogPriority.INFO)


    def execute(self):
        self.boundary_check()
        if self.last_send_time is None or time.time() - self.last_send_time > self.send_interval:
            self.send_sensor_data()
            self.last_send_time = time.time()
#            print("Sending sensor data", time.time())

File: src/modules/control_tasks/SensorControl.cpp
//
// Created by adiv413 on 4/24/2020.
//

#include <Logger/logger_util.h>
#include <flight/modules/control_tasks/SensorControl.hpp>
#include <flight/modules/lib/Enums.hpp>
#include <flight/modules/mcl/Config.hpp>
#include <chrono>

SensorControl::SensorControl() {
    this->last_send_time = 0;
    global_flag.log_info("response", {
        {"header", "info"},
        {"Description", "Sensor control started"}
    });
}

void SensorControl::begin() {
    log("Sensor control: Beginning");

    // Initialize the Kalman filters

    /* Pair of <string, <string, SensorInfo>> */
    for (const auto& type_ : global_config.sensors.list) {
        /* Pair of <string, SensorInfo> */
        for (const auto& location_ : type_.second) {
            ConfigSensorInfo sensor = location_.second;
            auto kalman = sensor.kalman_args;
            // Use brackets for the first because we want to create a new map
            // We use emplace for the second because Kalman has no default constructor
            kalman_filters[type_.first].emplace(location_.first, Kalman(
                kalman.process_variance,
                kalman.measurement_variance,
                kalman.kalman_value
            ));
        }
    }
}

void SensorControl::execute() {
    boundary_check();

    auto now = chrono::system_clock::now().time_since_epoch().count();

    if(last_send_time == 0 || now > last_send_time + global_config.sensors.send_interval) {
        send_sensor_data();
        last_send_time = now;
    }
}

bool between(double a, double b, double x) {
    return a <= x && x <= b;
}

void SensorControl::boundary_check() {
    vector<pair<string, string>> critical_sensors;

    for (const auto& type_ : global_config.sensors.list) {
        string type = type_.first;
        for (const auto& location_ : type_.second) {
            string location = location_.first;
            ConfigSensorInfo conf = location_.second;

            RegistrySensorInfo &sensor_registry = global_registry.sensors[type][location];
            double value = sensor_registry.measured_value;
            // We use .at() here because Kalman has no default constructor
            double kalman_value = kalman_filters.at(type).at(location).update_kalman(value);
            sensor_registry.normalized_value = kalman_value; // reference

            if (between(conf.boundaries.safe.lower, conf.boundaries.safe.upper, kalman_value)) {
                sensor_registry.status = SensorStatus::SAFE;
            } else if (between(conf.boundaries.warn.lower, conf.boundaries.warn.upper, kalman_value)) {
                sensor_registry.status = SensorStatus::WARNING;
            } else {
                sensor_registry.status = SensorStatus::CRITICAL;
                critical_sensors.push_back({type, location});
            }
        }
    }

    if (!global_registry.general.soft_abort and critical_sensors.empty()) { // one or more of the sensors are critical, soft abort
        global_registry.general.soft_abort = true;

        string message = "Soft aborting because the following sensors have reached critical levels- ";
        for(const pair<string, string>& sensor_location : critical_sensors) {
            message += sensor_location.first + "." + sensor_location.second + ", ";
        }
        message = message.substr(0, message.length() - 2);

        global_flag.log_critical("response", {
            {"header", "info"},
            {"Description", message}
        });
    }
}

void SensorControl::send_sensor_data() {
    json sensor_data_json = json::object();

    for (const auto& type_pair : global_config.sensors.list) {
        string type = type_pair.first;
        for (const auto &location_pair : type_pair.second) {
            string location = location_pair.first;
            RegistrySensorInfo sensor = global_registry.sensors[type][location];

            // "type.location": {
            sensor_data_json[type + "." + location] = json{
                {"measured", sensor.measured_value},
                {"kalman", sensor.normalized_value},
                {"get_status", int(sensor.status)}
            };
        }
    }

    global_flag.log_info("sensor_data", sensor_data_json);
}

